wal/src/message_queue_impl/region.rs

```diff
@@ -0,0 +1,761 @@
+        // just need to sync the snapshot while dropping table, but now we sync while
+        // every flushing... Just sync here now, obviously it is not enough.
+        synchronizer.sync(snapshot).await.context(SyncSnapshot)
+    }
+
+    /// See [RegionInner]'s related method.
+    pub async fn get_table_meta(&self, table_id: TableId) -> Result<TableMetaData> {
+        let inner = self.inner.read().await;
+        inner.get_table_meta(table_id).await
+    }
+
+    /// Clean outdated logs according to the information in region snapshot.
+    pub async fn clean_logs(&self) -> Result<()> {
+        // Get current snapshot.
+        let (snapshot, synchronizer) = {
+            let inner = self.inner.write().await;
+            (
+                inner.make_meta_snapshot().await,
+                self.snapshot_synchronizer.lock().await,
+            )
+        };
+
+        // Check and maybe clean logs.
+        self.log_cleaner
+            .maybe_clean_logs(&snapshot)
+            .await
+            .context(CleanLogs)?;
+
+        // Sync snapshot.
+        synchronizer.sync(snapshot).await.context(CleanLogs)
+    }
+
+    /// Return snapshot, just used for test.
+    async fn make_meta_snapshot(&self) -> RegionMetaSnapshot {
+        let inner = self.inner.write().await;
+        inner.make_meta_snapshot().await
+    }
+}
+
+/// Region's inner, all methods of [Region] are mainly implemented in it.
+#[allow(unused)]
+struct RegionInner<Mq> {
+    /// Id of region
+    region_id: RegionId,
+    /// Region meta data(such as, tables' next sequence numbers)
+    region_meta: RegionMeta,
+    /// Used to encode/decode the logs
+    log_encoding: CommonLogEncoding,
+    /// Message queue's Client
+    message_queue: Arc<Mq>,
+    /// Topic storing logs in message queue
+    log_topic: String,
+}
+
+#[allow(unused)]
+impl<Mq: MessageQueue> RegionInner<Mq> {
+    pub fn new(
+        region_id: RegionId,
+        log_encoding: CommonLogEncoding,
+        message_queue: Arc<Mq>,
+        log_topic: String,
+    ) -> Self {
+        // TODO: use snapshot to recover `region_meta`.
+        Self {
+            region_id,
+            region_meta: RegionMeta::default(),
+            log_encoding,
+            message_queue,
+            log_topic,
+        }
+    }
+
+    /// Write logs of table.
+    async fn write(
+        &self,
+        _ctx: &manager::WriteContext,
+        log_batch: &LogWriteBatch,
+    ) -> Result<SequenceNumber> {
+        // TODO: make use of ctx.
+        debug!(
+            "Begin to write to wal region, ctx:{:?}, region_id:{}, location:{:?}, log_entries_num:{}",
+            _ctx,
+            self.region_id,
+            log_batch.location,
+            log_batch.entries.len()
+        );
+        let location = &log_batch.location;
+        let log_write_entries = &log_batch.entries;
+
+        // Create messages and prepare for write.
+        let mut sequence_num = self
```

Comment:
Maybe we should name this `sequence_num` as `next_seq_num`, so in the end of this method it is natural to  return `next_seq_num - 1`.